pallets/rewards/src/lib.rs

```diff
@@ -0,0 +1,546 @@
+					T::NativeCurrency::set_lock(
+						REWARDS_LOCK_ID,
+						&user,
+						user_reward_info.total_reward_amount.saturated_into(),
+						WithdrawReasons::TRANSFER,
+					);
+					user_reward_info.is_initialized = true;
+					Ok(())
+				} else {
+					//sanity check
+					Err(Error::<T>::UserNotEligible)
+				}
+			})?;
+			Self::deposit_event(Event::UserUnlockedReward { user, reward_id });
+			Ok(())
+		}
+
+		/// The user will use the extrinsic to claim rewards
+		/// origin: The users address which has been mapped to reward id
+		/// id: The reward id
+		#[pallet::weight(10_000)]
+		pub fn claim(origin: OriginFor<T>, reward_id: u32) -> DispatchResult {
+			let user: T::AccountId = ensure_signed(origin)?;
+
+			//check if given id valid or not
+			ensure!(
+				<InitializeRewards<T>>::contains_key(reward_id),
+				Error::<T>::RewardIdNotRegister
+			);
+
+			//check if user is added in reward list
+			ensure!(<Distributor<T>>::contains_key(reward_id, &user), Error::<T>::UserNotEligible);
+
+			<Distributor<T>>::mutate(reward_id, user.clone(), |user_reward_info| {
+				if let Some(reward_info) = <InitializeRewards<T>>::get(reward_id) {
+					if let Some(user_reward_info) = user_reward_info {
+						//check if user has initialize rewards or not
+						ensure!(
+							user_reward_info.is_initialized,
+							Error::<T>::UserHasNotIntializeClaimRewards
+						);
+
+						//ensure that all rewards are not already claimed
+						ensure!(
+							user_reward_info.claim_amount.saturated_into::<u128>() !=
+								user_reward_info.total_reward_amount.saturated_into::<u128>(),
+							Error::<T>::AllRewardsAlreadyClaimed
+						);
+
+						let mut rewards_claimable: u128 = 0_u128.saturated_into();
+
+						//calculate the intial rewards that can be claimed
+						let initial_rewards_claimed = user_reward_info
+							.total_reward_amount
+							.saturated_into::<u128>()
+							.saturating_mul(reward_info.initial_percentage as u128)
+							.saturating_div(100);
+
+						//if intial rewards are not claimed add it to claimable rewards
+						if !user_reward_info.is_initial_rewards_claimed {
+							rewards_claimable = initial_rewards_claimed;
+						}
+
+						//calculate the number of blocks the user can claim rewards
+						let current_block_no: u128 =
+							<frame_system::Pallet<T>>::block_number().saturated_into();
+						let last_reward_claimed_block_no: u128 =
+							user_reward_info.last_block_rewards_claim.saturated_into();
+						let unclaimed_blocks: u128 =
+							min(current_block_no, reward_info.end_block.saturated_into::<u128>())
+								.saturating_sub(last_reward_claimed_block_no);
+						let crowdloan_period = reward_info
+							.end_block
+							.saturated_into::<u128>()
+							.saturating_sub(reward_info.start_block.saturated_into::<u128>());
+
+						//calculate custom factor for the user
+						// Formula = (total_rewards - intial_rewards_claimed) / crowloan_period
+						let factor = user_reward_info
```

Comment:
Can this factor be calculated and stored on the user_reward_info while initializing, so it only needs to be calculated once instead of on every claim?

Reply:
Thank you for implementing.